### PR TITLE
Fix issue with feature views being detected as duplicated when imported

### LINKED ISSUE #1895: FeatureView is considered "duplicated" during `feast apply` if its imported in another file
## Expected Behavior 
If I create a FeatureService in its own file, I need to import the desired FeatureViews. By doing so, `feat apply` is going to consider that `feature_view_a` is duplicated

## Current Behavior
If a FeatureView is imported in another file, `feast_apply` should not consider it as a duplicated FeatureView

Today, the current error is
```bash
ValueError: More than one feature view with name customer_raw_creation_info found.
Please ensure that all feature view names are unique.
It may be necessary to ignore certain files in your feature repository by using a .feastignore file.
```

## Steps to reproduce
1. Create a FeatureView name `feature_view_a` in its own file
2. Create a FeatureService in its own file and import `feature_view_a`
3. Run `feast apply`

### Specifications

- Version: 0.12.1
- Platform: MacOS bigSur
- Subsystem:

### PR DESCRIPTION
**Pull request by achals**
Signed-off-by: Achal Shah <achals@gmail.com>

**What this PR does / why we need it**:

**Which issue(s) this PR fixes**:

Fixes #1895 

**Does this PR introduce a user-facing change?**:

```release-note
none
```

### CODE REVIEW

File: sdk/python/feast/repo_operations.py [hunk 1]
```diff
@@ -107,15 +107,20 @@ def parse_repo(repo_root: Path) -> ParsedRepo:
         for attr_name in dir(module):
             obj = getattr(module, attr_name)
             if isinstance(obj, FeatureTable):
-                res.feature_tables.append(obj)
+                if obj.defined_in is not None and obj.defined_in == module.__file__:
```
Comment (MattDelac) [inline sdk/python/feast/repo_operations.py:110]: Can module.__file__ be None ?

File: sdk/python/feast/entity.py [hunk 2]
```diff
@@ -75,6 +77,11 @@ def __init__(
         self._created_timestamp: Optional[datetime] = None
         self._last_updated_timestamp: Optional[datetime] = None
 
+        stack = inspect.stack()
+        # Get two levels up from current, to ignore usage.py
+        previous_stack_frame = stack[2]
+        self.defined_in = previous_stack_frame.filename
```
Comment (MattDelac) [inline sdk/python/feast/entity.py:83]: Should we abstract that in an helper function and also test it in a test file ?
Don't think we need to test each class but at least the helper function

What do you think ?

**Review by MattDelac — COMMENTED** [review]
Nice !
I tried couple of the inspect functions like `getsourcefile()` in the other PR but could not make it work.
It makes sense to track where the objects are created 👍

File: sdk/python/feast/infra/online_stores/redis.py [hunk 3]
```diff
@@ -188,6 +188,7 @@ def online_read(
 
         for entity_key in entity_keys:
             redis_key_bin = _redis_key(project, entity_key)
+            print(redis_key_bin)
```
Comment (woop) [inline sdk/python/feast/infra/online_stores/redis.py:191]: should this be removed?

File: sdk/python/feast/repo_operations.py (hunk 1 above)
Comment (woop) [reply sdk/python/feast/repo_operations.py:110]: we should maybe also raise exceptions in case defined_in is None

**Comment by woop** [comment]
Just to play devil's advocate here. Tracking the file names does add complexity to our objects. Is there any reason we cant import all the objects and use built in functionality `__hash__()` and `__eq__` to ensure that they are the same if their names are the same, then we can dedupe/remove the additional ones?

File: sdk/python/feast/infra/online_stores/redis.py (hunk 3 above)
Comment (achals) [reply sdk/python/feast/infra/online_stores/redis.py:191]: Yes

**Comment by achals** [comment]
> Just to play devil's advocate here. Tracking the file names does add complexity to our objects. Is there any reason we cant import all the objects and use built in functionality `__hash__()` and `__eq__` to ensure that they are the same if their names are the same, then we can dedupe/remove the additional ones?

IMO updating `__hash__` and `__eq__` is more fragile since we may add/remove fields without updating those methods, and end up in problematic situations. I think the source file is a more robust solution and less likely to break with changes.

File: sdk/python/feast/entity.py (hunk 2 above)
Comment (adchia) [reply sdk/python/feast/entity.py:83]: A helper class probably makes sense since there's some state tracking being done. wdyt @achals ?

Comment (MattDelac) [reply sdk/python/feast/entity.py:83]: > A helper class

yeah even better 👍

Comment (achals) [reply sdk/python/feast/entity.py]: Done

File: sdk/python/feast/feature_table.py [hunk 4]
```diff
@@ -64,6 +66,11 @@ def __init__(
         self._created_timestamp: Optional[Timestamp] = None
         self._last_updated_timestamp: Optional[Timestamp] = None
 
+        stack = inspect.stack()
+        # Get two levels up from current, to ignore usage.py
+        previous_stack_frame = stack[2]
+        self.defined_in = previous_stack_frame.filename
```
Comment (MattDelac) [inline sdk/python/feast/feature_table.py:72]: Sounds like missing one `get_calling_file_name()` here

File: sdk/python/feast/infra/online_stores/redis.py (hunk 3 above)
Comment (MattDelac) [inline sdk/python/feast/infra/online_stores/redis.py:191]: To be removed ?

File: sdk/python/feast/feature_table.py (hunk 4 above)
Comment (achals) [reply sdk/python/feast/feature_table.py:72]: oh yeah missed this

File: sdk/python/feast/repo_operations.py (hunk 1 above)
Comment (achals) [reply sdk/python/feast/repo_operations.py]: Done. And no, module.__file__ cannot be None

**Review by adchia — APPROVED** [review]
/lgtm

**Comment by MattDelac** [comment]
👋 
I just tried it on our repo and we shipped with a bug. The problem is that we save in `defined_in` one level too high.

Look at how my `feast apply` behaves
image

`module.__file__` is correct but `obj.defined_in` is not.
The entity "customer" is, indeed, defined in `pano/entities/customer.py`

**Comment by woop** [comment]
Whether or not there is a bug here, I am not 100% convinced by the approach taken here. There are many reasons to support object equality, deduplication just being one of them. Reaching into the filesystem seems like it will be a lot more prone to errors since its OS/env dependent.

**Comment by achals** [comment]
> Whether or not there is a bug here, I am not 100% convinced by the approach taken here. There are many reasons to support object equality, deduplication just being one of them. Reaching into the filesystem seems like it will be a lot more prone to errors since its OS/env dependent.

I'm coming around to this view point after this bug as well. It's pretty easy to get wrong. I'm going to undo this change and use equality based deduping.

**Pull request merged**